adam/learner/integrated_learner.py

```diff
@@ -0,0 +1,395 @@
+            self.relation_learner,
+        ]:
+            if sub_learner:
+                cur_description_state = sub_learner.enrich_during_description(
+                    cur_description_state
+                )
+
+        if perception.is_dynamic() and self.action_learner:
+            cur_description_state = self.action_learner.enrich_during_description(
+                cur_description_state
+            )
+
+        return self._linguistic_descriptions_from_semantics(
+            cur_description_state.semantic_nodes
+        )
+
+    def _linguistic_descriptions_from_semantics(
+        self, semantic_nodes: AbstractSet[SemanticNode]
+    ) -> Mapping[LinguisticDescription, float]:
+        learner_semantics = LearnerSemantics.from_nodes(semantic_nodes)
+
+        ret = []
+        if self.action_learner:
+            ret.extend(
+                [
+                    (action_tokens, 1.0)
+                    for action in learner_semantics.actions
+                    for action_tokens in self._instantiate_action(
+                        action, learner_semantics
+                    )
+                    # ensure we have some way of expressing this action
+                    if self.action_learner.templates_for_concept(action.concept)
+                ]
+            )
+
+        if self.relation_learner:
+            ret.extend(
+                [
+                    (relation_tokens, 1.0)
+                    for relation in learner_semantics.relations
+                    for relation_tokens in self._instantiate_relation(
+                        relation, learner_semantics
+                    )
+                    # ensure we have some way of expressing this relation
+                    if self.relation_learner.templates_for_concept(relation.concept)
+                ]
+            )
+
+        ret.extend(
+            [
+                (object_tokens, 1.0)
+                for object_ in learner_semantics.objects
+                for object_tokens in self._instantiate_object(object_, learner_semantics)
+                # ensure we have some way of expressing this object
+                if self.object_learner.templates_for_concept(object_.concept)
+            ]
+        )
+
+        return immutabledict(
+            (TokenSequenceLinguisticDescription(tokens), score) for (tokens, score) in ret
+        )
+
+    def _instantiate_object(
+        self, object_node: ObjectSemanticNode, learner_semantics: "LearnerSemantics"
+    ) -> Iterator[Tuple[str, ...]]:
+        # For now, we assume the order in which modifiers is expressed is arbitrary.
+        attributes_we_can_express = (
+            [
+                attribute
+                for attribute in learner_semantics.objects_to_attributes[object_node]
+                if self.attribute_learner.templates_for_concept(attribute.concept)
+            ]
+            if self.attribute_learner
+            else []
+        )
+        # TODO: deal with relations
```

Comment:
@spigo900 : Please make and reference an issue about how we can express "top-level" relations, but we can't yet express relations as modifiers of objects embedded in action descriptions, etc.

Reply:
@gabbard Issue #794.